uat/custom-components/client-java-paho/src/main/java/com/aws/greengrass/testing/mqtt311/client/sdkmqtt/Mqtt311ConnectionImpl.java

```diff
@@ -0,0 +1,117 @@
+
+/**
+ * Implementation of MQTT 3.1.1 connection based on AWS IoT SDK.
+ */
+public class Mqtt311ConnectionImpl implements MqttConnection {
+    private static final Logger logger = LogManager.getLogger(Mqtt311ConnectionImpl.class);
+    private static final String EXCEPTION_WHEN_CONNECTING = "Exception occurred during connect";
+
+    private final AtomicBoolean isClosing = new AtomicBoolean();
+    private final AtomicBoolean isConnected = new AtomicBoolean();
+
+    private final GRPCClient grpcClient;
+    private final IMqttClient mqttClient;
+    private int connectionId = 0;
+
+    private final ExecutorService executorService = Executors.newSingleThreadExecutor();        // TODO: use DI
+
+    /**
+     * Creates a MQTT 3.1.1 connection.
+     *
+     * @param connectionParams the connection parameters
+     * @param grpcClient the consumer of received messages and disconnect events
+     * @throws MqttException on errors
+     */
+    public Mqtt311ConnectionImpl(@NonNull MqttLib.ConnectionParams connectionParams, @NonNull GRPCClient grpcClient)
+            throws org.eclipse.paho.client.mqttv3.MqttException {
+        super();
+        this.grpcClient = grpcClient;
+        this.mqttClient = createClient(connectionParams);
+    }
+
+    @Override
+    public ConnectResult start(long timeout, int connectionId) throws MqttException {
+        this.connectionId = connectionId;
+        try {
+            IMqttToken token = mqttClient.connectWithResult(new MqttConnectOptions());
+            token.waitForCompletion();
+            isConnected.set(true);
+            logger.atInfo().log("MQTT 3.1.1 connection {} is establisted", connectionId);
+            return buildConnectResult(true, token.isComplete());
+        } catch (org.eclipse.paho.client.mqttv3.MqttException e) {
+            logger.atError().withThrowable(e).log(EXCEPTION_WHEN_CONNECTING);
+            throw new MqttException(EXCEPTION_WHEN_CONNECTING, e);
+        }
+    }
+
+    @Override
+    public void disconnect(long timeout, int reasonCode) throws MqttException, org.eclipse.paho.client.mqttv3.MqttException {
+
+        if (!isClosing.getAndSet(true)) {
+            try {
+                mqttClient.disconnect();
+
+                logger.atInfo().log("MQTT 3.1.1 connection {} has been disconnected", connectionId);
+            } catch (org.eclipse.paho.client.mqttv3.MqttException e) {
+                throw new RuntimeException(e);
+            } finally {
+                mqttClient.close();
+            }
+        }
+    }
+
+    /**
+     * Creates a MQTT 311 connection.
+     *
+     * @param connectionParams connection parameters
+     * @return MQTT 3.1.1 connection
+     * @throws MqttException on errors
+     */
+    private IMqttClient createClient(MqttLib.ConnectionParams connectionParams) throws org.eclipse.paho.client.mqttv3.MqttException {
+        IMqttClient client = new MqttClient(connectionParams.getHost(), connectionParams.getClientId());
```

Comment:
For future connect tasks: please store ALL useful members of connectionParams like TLS credentials